### PR TITLE
Parse type annotations on method parameters and return types (BT-585)

### PR DESCRIPTION
**Pull request by jamesc**
## Summary

Implements ADR 0025 Phase 2 syntax parsing: optional type annotations on method parameters and return types.

**Linear issue:** https://linear.app/beamtalk/issue/BT-585

### AST
- Added `ParameterDefinition` struct with `name: Identifier` + `type_annotation: Option<TypeAnnotation>`
- Changed `MethodDefinition.parameters` from `Vec<Identifier>` to `Vec<ParameterDefinition>`

### Parser
- Enhanced `parse_type_annotation()` to handle union types (`Integer | String`)
- Added `parse_optional_return_type()` for `-> Type` syntax before `=>`
- Updated keyword method parsing for typed params (`deposit: amount: Integer =>`)
- Updated binary method parsing for typed params (`+ other: Number =>`)
- Updated lookahead functions (`is_at_method_definition`, `is_keyword_method_at`) for new patterns

### Type Checker
- Extracted `set_param_types()` helper to avoid code duplication
- Simple type annotations (e.g., `Integer`) set as Known types in TypeEnv
- Union/Generic annotations stored but not yet wired (Phase 3 per ADR 0025)

### Downstream Consumers
- Updated all `.parameters` consumers: codegen (actor, gen_server, value_type, doc_chunks), semantic analysis (mod.rs, name_resolver), CLI doc command

### Tests
- 9 new parser tests covering: return types, union types, typed keyword/binary params, mixed typed/untyped
- 7 snapshot tests updated for new `ParameterDefinition` AST shape
- All 1381 stdlib + 1329 runtime + E2E tests pass

## Syntax Examples

```beamtalk
// Return type
increment -> Integer => count := count + 1

// Typed keyword parameter
deposit: amount: Integer => balance := balance + amount

// Multiple typed params with return type
transfer: amount: Integer to: target: BankAccount -> Boolean =>
  // ...

// Binary method with typed param
+ other: Number -> Number => self + other

// Union return type
find: key: String -> Integer | Nil => // ...
```

## Summary by CodeRabbit

* **New Features**
  * Methods can declare typed parameters with optional `: Type` annotations.
  * Methods can declare return types using `-> Type`, including union forms like `Type | Type`.
  * Simple parameter type annotations are now used by static analysis to better recognize parameter types and improve checks.

* **Tests**
  * Expanded coverage for typed parameters and return-type parsing across unary, binary, and keyword methods, plus many recovery/edge-case scenarios.

### CODE REVIEW

File: crates/beamtalk-core/src/source_analysis/parser/declarations.rs [hunk 1]
```diff
@@ -548,8 +662,43 @@ impl Parser {
...
+                        }
+                    }
+
+                    // Parse untyped parameter name
+                    let param_name = self.parse_identifier("Expected parameter name after keyword");
+                    let type_annotation = self.parse_optional_param_type();
+                    let param = match type_annotation {
+                        Some(ta) => ParameterDefinition::with_type(param_name, ta),
+                        None => ParameterDefinition::new(param_name),
+                    };
```
[the remark this answers is not in the document]
Comment (jamesc) [reply crates/beamtalk-core/src/source_analysis/parser/declarations.rs]: Fixed in 67e26d5. Replaced the uppercase heuristic with token-based disambiguation: after `paramName:` + `Identifier`, check if the next token is `=>`, `->`, or another `Keyword` — if so, the Identifier is a type name (typed parameter). This is consistent with `is_keyword_method_at` and correctly handles lowercase type names/type variables.

File: crates/beamtalk-core/src/source_analysis/parser/declarations.rs [hunk 2]
```diff
@@ -238,27 +238,65 @@ impl Parser {
...
+        // Skip union types: `| Type`
+        while matches!(self.peek_at(o), Some(TokenKind::Pipe)) {
+            o += 1;
+            if !matches!(self.peek_at(o), Some(TokenKind::Identifier(_))) {
+                return false;
+            }
+            o += 1;
+        }
+        matches!(self.peek_at(o), Some(TokenKind::FatArrow))
+    }
```
[the remark this answers is not in the document]
Comment (jamesc) [reply crates/beamtalk-core/src/source_analysis/parser/declarations.rs]: Fixed in 26d550a. Added early `-> =>` (missing type) acceptance in `is_return_type_then_fat_arrow` so the parser detects the method definition and lets `parse_type_annotation` emit "Expected type name" error rather than truncating class parsing. Added test `parse_malformed_return_type_recovers` to verify the method is still found with a diagnostic.
